File: examples/uid_lookup_example.cpp
#include <iomanip>
#include <iostream>
#include <string_view>

#include <dicom.h>

namespace {

void print_usage(const char* argv0) {
    std::cout << "Usage: " << argv0 << " <uid-or-keyword> [<uid-or-keyword> ...]\n";
    using namespace dicom::literals;
    constexpr dicom::Uid implicit = "ImplicitVRLittleEndian"_uid;
    std::cout << "Example literal: \"ImplicitVRLittleEndian\"_uid -> "
              << implicit.value() << " (" << implicit.keyword() << ")\n";
}

}  // namespace

int main(int argc, char** argv) {
    if (argc < 2 || std::string_view(argv[1]) == "--help" || std::string_view(argv[1]) == "-h") {
        print_usage(argv[0]);
        return argc < 2 ? 1 : 0;
    }

    for (int i = 1; i < argc; ++i) {
        const std::string_view text{argv[i]};
        const dicom::Uid uid = dicom::Uid::lookup(text);
        if (!uid) {
            std::cout << text << ": not found\n";
            continue;
        }
        std::cout << text << " -> value=" << uid.value();
        if (!uid.keyword().empty()) {
            std::cout << ", keyword=" << uid.keyword();
        }
        std::cout << ", name=" << uid.name() << ", type=" << uid.type() << "\n";
    }
    return 0;
}

File: examples/python/uid_lookup_example.py
#!/usr/bin/env python3
"""Lookup well-known DICOM UIDs using the dicomsdl Python bindings."""

from __future__ import annotations

import argparse
import sys

import dicomsdl as dicom


def describe(uid: dicom.Uid, original: str) -> str:
    keyword = uid.keyword
    keyword_str = f", keyword='{keyword}'" if keyword else ""
    return (
        f"{original} -> value='{uid.value}'{keyword_str}, "
        f"name='{uid.name}', type='{uid.type}'"
    )


def main(argv: list[str]) -> int:
    parser = argparse.ArgumentParser(
        description="Lookup UIDs by dotted value or keyword using dicomsdl",
    )
    parser.add_argument("uids", nargs="+", help="UID dotted values or keywords")
    args = parser.parse_args(argv)

    for text in args.uids:
        uid = dicom.lookup_uid(text)
        if uid is None:
            print(f"{text}: not found")
            continue
        print(describe(uid, text))
    return 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv[1:]))
